### PR TITLE
PACA-275: RDA claim parallel writes

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-275](https://jira.cms.gov/browse/PACA-275)

**User Story or Bug Summary:**

Modify the RDA API pipeline to improve claim ingestion performance.

---

### What Does This PR Do?

Thie PR implements several changes to increase claim processing throughput:

- Modifies the `RdaSink` interface to allow it to be implemented asynchronously.
- Modifies `GrpcRdaSource` to use the new `RdaSink` methods when saving data to the database.
- Adds a new table `RdaApiProgress` to store sequence numbers for each claim type.  These are used to know what starting sequence number to request when calling the RDA API.  During data ingestion the `WriterThreadPool` class uses a `SequenceNumberTracker` class to determine which sequence number should be written to this table and the `SequenceNumberWriterThread` does the actual writing.  Comments in these classes explain the logic.
- Moves `message->claim` transformation into the `RdaSink` so that it can be done in worker threads.  Previously this transformation was done in the `GrpcStreamCaller` classes which was a bottleneck when using asynchronous writes.
- Adds a new `ConcurrentRdaSink` that uses a pool of worker threads to write claims and sequence numbers to the database asynchronously.  This is a general wrapper that delegates work to entity specific single-threaded sink implementations.  During data ingestion the `ConcurrentRdaSink` uses a `WriterThreadPool` to manage a pool of multiple `ClaimWriterThread`s and a single `SequenceNumberWriterThread` to perform the database updates.
- Modifies `AppConfiguration` to allow the number of claim writer threads to be configured for each job using an environment variable.  Note that this controls the number of writers per job.  Since there are two jobs (one for FISS and another for MCS claims) the actual number of writer threads will be double this number.
- Splits the `RdaSink` implementation classes into two new packages (`gov.cms.bfd.pipeline.rda.grpc.sink.direct` and `gov.cms.bfd.pipeline.rda.grpc.sink.concurrent`) to make the code easier to follow.  **This refactoring created some noise in the `Files changed` list but was worthwhile since the code in the two packages are logically very different.**. The `direct` classes have all the knowledge about FISS/MCS claims while the `concurrent` classes have no knowledge of those but instead have knowledge of threads/queues etc.
- Removes `persist() then merge()` optimization logic from `AbstractClaimRdaSink`.  That feature added complexity and was tuned for incoming data that was almost all new claims.  Current understanding is that many or most claims will be updates to existing records so the extra complexity wasn't worthwhile.  We can revisit this once we have prod data history next year.
- Adds a `CachingIdHasher` to avoid recomputing recently computed MBI values.  For now uses an in memory cache but can easily be extended to perform database lookups in a future PR.
- Allows `RdaServer` to be configured to listen on a specific host name for added security when benchmarking.

### How do the pieces fit together?

The pipeline job defines an interface named `RdaSink` for objects that handle data transformation and storage of the objects it ingests.  Objects that implement `RdaSink` do the following:

- Read and write sequence numbers.  More on these below.
- Transform RDA API change messages to JPA entities.
- Write JPA change entities to the database.

The RDA API returns a stream of change messages.  Each change message has a sequence number and a claim.  The sequence numbers are unique to each change and increase with every change.  Callers can send a starting sequence number to resume at that point in the stream.  We store only the latest version of each claim in our database.

There are 3 `RdaSink` implementations:

- `FissClaimRdaSink` synchronous implementation that supports FISS claims.
- `McsClaimRdaSink` synchronous implementation that supports MCS claims.
- `ConcurrentRdaSink` asynchronous implementation that is a wrapper on top of one of the other two.

The `ConcurrentRdaSink`'s worker threads use the appropriate sink implementation to perform the actual transformation and database writes.  Every incoming claim is assigned to a worker thread based on a hash of its claim ID.  This is done to ensure that we can't write two changes for the same claim out of sequence.

As claims are scheduled for writing we have to track their sequence numbers so that we can maintain the correct value in the `RdaApiProgress` table.  The correct value is always the highest sequence number for which we know there are no unsaved changes with a lower sequence number.   This logic is handled using a class called `SequenceNumberTracker`.

There are two types of worker threads.
- `SequenceNumberWriterThread` updates the `RdaApiProgress` table.  There is always one of these for each pool.
- `ClaimWriterThread` updates the appropriate claim tables.  There are a variable number of these based on configuration settings.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:
* Verify all PR security questions and checklists have been completed and addressed.
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* This PR contains a schema change that appeared to require a separate PR to be merged ahead of it to disable hibernate schema validation.
* [PR 904](https://github.com/CMSgov/beneficiary-fhir-data/pull/904)

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java:140]: Description here for throws

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:92]: This seems like it could possibly be a non-trivial calculation, should we wrap this with a `logger.infoEnabled()` check?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:117]: This seems like a very complex method, curious if this can be broken up a little for better readability

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaSink.java
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaSink.java:59]: Update javadocs for params

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java:62]: This is a command line program used for local testing so it doesn't use logging for basic output.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java:100]: Since this is just run for local testing/benchmarking that's not a problem.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/assembly/zip.xml
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/assembly/zip.xml:1]: I need to be able to upload the server jar to an EC2 instance to run it for benchmarking.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java:37]: Good point.  Are we using`@Slf4j` in other places?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java:28]: Absolutely necessary here to ensure claims are written on a single thread.  The failure mode here is simple.  If any thread fails the whole process stops (just as a single threaded sink would work).

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:92]: Since answer is an optional this isn't really expensive.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java
Comment (KyRobbins) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java:37]: I'm using it in my bridge CLI tool, but I'm not aware of it being used in any other module, as we don't have lombok in anything except the new modules we added I don't think.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java
Comment (KyRobbins) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ClaimWriterThread.java:28]: I actually realized this later after writing the comment when I saw notes about maintaining order, chronographically, of like claims, but thanks for the verification!

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:117]: It's all very linear.  It hasn't changed much from master and since it's mostly catch blocks to carry through the correct processed count in a ProcessingException it can't be separated out much.  I tried and the separate method just contains the about the same set of catch blocks as its caller.  I suggest we consider that as a refactoring of its own for a future ticket.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberWriterThread.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberWriterThread.java:164]: Nothing.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java]: Fixed.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__api_progress_table.sql
Comment (ericbuckley) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__api_progress_table.sql:9]: with this addition, do we need to keep indexes for the sequence numbers in the FISS and MCS claims tables?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java:72]: I'm pretty sure this doesn't matter, but I think there is a chance here of returning a value that has yet to be recorded in the DB.  There is no guarantee that we will receive a message for every possible sequence number value.  If for example RDA never sends a message for sequence number 100, but at the time of running this function the smallest number in the set for 101, this function will return 100.  Again, this seems fine if all we need is a number to know where to continue from, but if there is some logic that assumes this value DOES exist in our DB we might have a problem.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:86]: any reason to not using a parameter for the claimType?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java
Comment (brianburtonadhoc) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java:61]: This is a command line utility so direct output to standard out is helpful.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaLoadOptions.java
Comment (brianburtonadhoc) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaLoadOptions.java:26]: Yes I support that 100%

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__api_progress_table.sql
Comment (brianburtonadhoc) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__api_progress_table.sql:9]: Good catch.  I will update the migration script to drop those indexes.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java
Comment (brianburtonadhoc) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java:72]: Your analysis is correct.  The purpose of this is to track safe values for the sequence number column in the progress table.  It's not to guarantee that the number is always a value received from the API but it will be a value between the value in the database when the pipeline started and the lowest unwritten value that we have received while the pipeline was running.

Any logic that depends on the value in this column actually existing in a record within the claims tables would be a pretty serious design bug.  There is no integrity constraint relating to this table to records in the claim tables.  Also there is aonly a single column holding a sequence number so no good way to create any such constraint.  No cxonstraint is needed though because this table is really just a scratch pad to allow the pipeline to track its best known resumption point as it runs and between runs.

Even if someone tried to work out a way to have a constraint like that it would be doomed to cause problems simply because the transaction that writes the sequence numbers to the progress table is decoupled from the transactions that update the claims themselves.  It's not unlikely that the sequence number value to be written at this moment might already have been made obsolete by another transaction updating the claim that originally had that sequence number.  There is no harm in that since in the worst case of the pipeline crashing at that moment and restarting it would just resume at the older point and reprocess something it had already done but not recorded in the progress table yet.

That's by design since it safely allows all of the claim update transactions to work within just the claims tables.  They cannot conflict with each other while trying to also update the same record of the progress table in the same transaction.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java
Comment (brianburtonadhoc) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:86]: Yes because concrete classes that derive from this are dedicated to a single claim type and the user of those classes does not need to know or care what the actual claim type is.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__api_progress_table.sql
Comment (brianburton) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V37__api_progress_table.sql]: FYI - the migration file had to be renamed following a merge because it had the same version number as a migration from the merge.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaLoadOptions.java
Comment (ericbuckley) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaLoadOptions.java:26]: 👍 I created [PACA-350](https://jira.cms.gov/browse/PACA-350)

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java
Comment (ericbuckley) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:86]: I meant at the query level.
```
      String query =
          String.format(
              "select p.%s from RdaApiProgress p where p.claimType=:claimType",
              RdaApiProgress.Fields.lastSequenceNumber);
      final List<Long> sequenceNumber =
          entityManager.createQuery(query, Long.class)
             .setParameter("claimType", claimType.name())
             .getResultList();
```
maybe I'm over thinking it

**Review by ericbuckley — APPROVED** [review]

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:86]: Ah I see.  Good point.  I think given this string is from a constant it's not a big deal to just add it directly.

**Review by ericbuckley — DISMISSED** [review]

**Review by KyRobbins — DISMISSED** [review]

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java
Comment (christrotteradhoc) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java:68]: In the context of the discussion below I would suggest renaming this method to 'getSafeResumeSequenceNumber' or something along those lines. The method name as it is currently here implies more than it should and future maintainers may make assumptions based on that.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java
Comment (christrotteradhoc) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java:52]: Can you elaborate some on your choice of a cached thread pool here as opposed to a fixed thread pool? Have we been able to load test this at production levels and ensure that the size of this pool was kept to reasonable levels? It might be useful to include somewhere in here as a comment as well the factors that led to choosing this.

**Review by christrotteradhoc — COMMENTED** [review]
This is really impressive. Nice work! I have a small comment inline and I am ready to approve once we are ready to do the multi-staged deployment next week.

**Review by christrotteradhoc — CHANGES_REQUESTED** [review]
Blocking so that we don't accidentally merge until we have split it up and followed the migration deployment process.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java:52]: Cached thread pool creates only as many threads as needed to execute submitted jobs and shuts down any unused/cached threads if they are idle very long.  I prefer them here because it's easy to get the count wrong when creating a fixed thread pool and have a stuck job that never launches or an unused thread that's sitting idle and wasting resources.  The count issue is particularly likely when a class is updated later to add a new type worker thread.

In this case the pool is created and all jobs are submitted at the beginning so there is no way for the pool to run away on us.  It will always have exactly one thread per job.  Which in this case is one thread per ClaimWriter or SequenceNumberWriter.

**Comment by ericbuckley** [comment]
> Blocking so that we don't accidentally merge until we have split it up and followed the migration deployment process.

@christrotteradhoc we spoke about this today and we think the migration + code changes are safe to be deployed at the same time.
1. This migration is backwards compatible, as adding a new table and removing indexes will NOT prevent old versions of the code base from running (removing the indexes would cause an older version of the API to run slower in production, but being that we don't have any data and our API isn't live, this doesn't really matter)
2. We tested the hibernate validation locally and it does not assert an error when unknown tables exist in the schema

What do you think about merging the migration with the code changes?

**Review by ericbuckley — APPROVED** [review]

**Review by christrotteradhoc — APPROVED** [review]

**Pull request merged**
